`Task :expo-constants:generateReleaseLintModel FAILED` when running `gradlew assembleRelease`

### Minimal reproducible example

https://github.com/achmadk/ExpoConstantErrorGenerateReleaseLintModel

### Summary

I got this error when running `cd android && gradlew assembleRelease`

```sh
> Task :expo-constants:generateReleaseLintModel FAILED

FAILURE: Build failed with an exception.

* What went wrong:                                                                                                  
A problem was found with the configuration of task ':expo-constants:generateReleaseLintModel' (type 'LintModelWriterTask').
  - Gradle detected a problem with the following location: 'D:\PROJECTS\SIDE_PROJECTS\ExpoConstantErrorGenerateReleaseLintModel\node_modules\expo-constants\android\src\release\assets'.
    
    Reason: Task ':expo-constants:generateReleaseLintModel' uses this output of task ':expo-constants:copyReleaseExpoConfig' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':expo-constants:copyReleaseExpoConfig' as an input of ':expo-constants:generateReleaseLintModel'.
      2. Declare an explicit dependency on ':expo-constants:copyReleaseExpoConfig' from ':expo-constants:generateReleaseLintModel' using Task#dependsOn.
      3. Declare an explicit dependency on ':expo-constants:copyReleaseExpoConfig' from ':expo-constants:generateReleaseLintModel' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/8.4/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.4/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 2m 15s
```

### Environment

```sh
 expo-env-info 1.0.5 environment info:
    System:
      OS: Windows 10 10.0.22631
    Binaries:
      Node: 20.10.0 - C:\Program Files\nodejs\node.EXE
      Yarn: 4.0.2 - D:\PROJECTS\SIDE_PROJECTS\wedew-guestbook-mobile\node_modules\.bin\yarn.CMD
      npm: 10.2.5 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: AI-231.9392.1.2311.11076708
    npmPackages:
      expo: ^50.0.0-preview.6 => 50.0.0-preview.6
      react: 18.2.0 => 18.2.0
      react-dom: 17.0.2 => 17.0.2
      react-native: 0.73.1 => 0.73.1
    Expo Workflow: bare
```

--- a/packages/expo-constants/CHANGELOG.md
+++ b/packages/expo-constants/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- Fixed build error on AGP 8.2. ([#26362](https://github.com/expo/expo/pull/26362) by [@kudo](https://github.com/kudo))
+
 ### 💡 Others
 
 - [expo-updates] Migrate to requireNativeModule/requireOptionalNativeModule. ([#25648](https://github.com/expo/expo/pull/25648) by [@wschurman](https://github.com/wschurman))

--- a/packages/expo-constants/scripts/get-app-config-android.gradle
+++ b/packages/expo-constants/scripts/get-app-config-android.gradle
@@ -10,63 +10,32 @@ def nodeExecutableAndArgs = config.nodeExecutableAndArgs ?: ["node"]
 
 afterEvaluate {
   def projectRoot = file("${rootProject.projectDir}")
+  def assetsDir = file("$buildDir/generated/assets/expo-constants")
 
-  if (!android.hasProperty('libraryVariants')) {
-    // For traditional application level integration, will be no-op to prevent duplicated app.config creation.
-    return;
+  def currentCreateConfigTask = tasks.register('createExpoConfig', Exec) {
+    description = 'expo-constants: Create app.config.'
+
+    doFirst {
+      assetsDir.deleteDir()
+      assetsDir.mkdirs()
+    }
+
+    // Add generated assetsDir into assets.srcDirs
+    project.android.sourceSets.main.assets.srcDirs += assetsDir
+
+    // Set up outputs so gradle can cache the result
+    outputs.dir assetsDir
+
+    // Switch to project root and generate the app config
+    workingDir projectRoot
+
+    if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+      commandLine("cmd", "/c", *nodeExecutableAndArgs, "$expoConstantsDir/scripts/getAppConfig.js", projectRoot, assetsDir)
+    } else {
+      commandLine(*nodeExecutableAndArgs, "$expoConstantsDir/scripts/getAppConfig.js", projectRoot, assetsDir)
+    }
   }
-  android.libraryVariants.each { variant ->
-    def targetName = variant.name.capitalize()
-    def targetPath = variant.dirName
 
-    def assetsDir = file("$buildDir/generated/assets/expo-constants/${targetPath}")
-    def packageTaskName = "package${targetName}Assets"
-
-    def currentBundleTask = tasks.register("create${targetName}ExpoConfig", Exec) {
-      description = "expo-constants: Create config for ${targetName}."
-
-      doFirst {
-        assetsDir.deleteDir()
-        assetsDir.mkdirs()
-      }
-
-      // Set up outputs so gradle can cache the result
-      outputs.dir assetsDir
-
-      // Switch to project root and generate the app config
-      workingDir projectRoot
-
-      if (Os.isFamily(Os.FAMILY_WINDOWS)) {
-        commandLine("cmd", "/c", *nodeExecutableAndArgs, "$expoConstantsDir/scripts/getAppConfig.js", projectRoot, assetsDir)
-      } else {
-        commandLine(*nodeExecutableAndArgs, "$expoConstantsDir/scripts/getAppConfig.js", projectRoot, assetsDir)
-      }
-    }
-
-    def currentCopyAppConfigTask = tasks.register("copy${targetName}ExpoConfig", Copy) {
-      description = "expo-constants: Copy generated app.config into ${targetName}."
-
-      from(assetsDir)
-      into("${projectDir}/src/${targetPath}/assets")
-      include('app.config')
-
-      dependsOn(currentBundleTask)
-    }
-
-    def currentCleanupAppConfigTask = tasks.register("cleanup${targetName}ExpoConfig", Delete) {
-      description = "expo-constants: Cleanup generated app.config from ${targetName}."
-
-      // In the past, the `app.config` file was placed in the `main` directory.
-      // However, this is no longer the case. To ensure a smooth transition,
-      // we’re also cleaning the `app.config` from the old location.
-      delete files("${projectDir}/src/main/assets/app.config")
-      delete files("${projectDir}/src/${targetPath}/assets/app.config")
-      dependsOn(packageTaskName)
-    }
-
-    // In summary, these tasks try to create `src/${targetPath}/assets/app.config` in building time for gradle to merge the asset.
-    // And cleanup the generated app.config after building time.
-    tasks.findByPath(packageTaskName).dependsOn(currentCopyAppConfigTask)
-    tasks.findByPath("process${targetName}JavaRes").dependsOn(currentCleanupAppConfigTask)
-  }
+  // Generate app.config at preBuild
+  tasks.getByName('preBuild').dependsOn(currentCreateConfigTask)
 }

